### PR TITLE
[Merged by Bors] - feat(CategoryTheory): wide subcategories

### PR DESCRIPTION
**Pull request by sinhp**
This PR introduces the notion of wide subcategory. The API is in parallel with that of full subcategories. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Widesubcategory.lean [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+universe v₁ v₂ u₁ u₂
+
+open MorphismProperty
+
+section Induced
+
+variable {C : Type u₁} (D : Type u₂) [Category.{v₁} D]
+variable (F : C → D) (P : MorphismProperty D)
+    (hP : IsMultiplicative P)
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Widesubcategory.lean:37]: `IsMultiplicative` is a type-class:
```suggestion
    [IsMultiplicative P]
```

Comment (sinhp) [reply Mathlib/CategoryTheory/Widesubcategory.lean:37]: great! This shortened `InducedWideCategory D F P hP` to `InducedWideCategory D F P` all throughout the file.

Comment (sinhp) [reply Mathlib/CategoryTheory/Widesubcategory.lean:37]: and shortened `WideSubcategory P hP` to `WideSubcategory P`.

File: Mathlib/CategoryTheory/Widesubcategory.lean [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+@[simps!]
+instance InducedWideCategory.category :
+    Category (InducedWideCategory D F P) where
+  Hom X Y := {f : F X ⟶ F Y | P f}
+  id X := ⟨𝟙 (F X), P.id_mem (F X)⟩
+  comp {X Y Z} f g := ⟨f.1 ≫ g.1, P.comp_mem _ _ f.2 g.2⟩
+
+/-- The forgetful functor from an induced wide category to the original category,
+forgetting the extra data.
+-/
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Widesubcategory.lean:62]: It does not seem to be that any data is forgotten.
```suggestion
/-- The forgetful functor from an induced wide category to the original category. -/
```

File: Mathlib/CategoryTheory/Widesubcategory.lean [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+/--
+Structure for wide subcategories. Objects ignore the morphism property.
+-/
+@[ext, nolint unusedArguments]
+structure WideSubcategory (_P : MorphismProperty C) [IsMultiplicative _P] where
+  /-- The category of which this is a wide subcategory-/
+  obj : C
+
+@[simps!]
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Widesubcategory.lean:89]: What about making `id_def` and `comp_def` simp lemmas instead?
```suggestion
```

File: Mathlib/CategoryTheory/Widesubcategory.lean [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  rfl
+
+@[simp]
+theorem wideSubcategoryInclusion.map {X Y} {f : X ⟶ Y} :
+    (wideSubcategoryInclusion P ).map f = f.1 :=
+  rfl
+
+/-- The inclusion of a wide subcategory is faithful. -/
+instance wideSubcategory.faithful : (wideSubcategoryInclusion P).Faithful := by
+  exact inferInstanceAs (wideInducedFunctor WideSubcategory.obj P).Faithful
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Widesubcategory.lean:115]:
```suggestion
instance wideSubcategory.faithful : (wideSubcategoryInclusion P).Faithful :=
  inferInstanceAs (wideInducedFunctor WideSubcategory.obj P).Faithful
```

File: Mathlib/CategoryTheory/Widesubcategory.lean [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+def wideSubcategoryInclusion : WideSubcategory P ⥤ C :=
+  wideInducedFunctor WideSubcategory.obj P
+
+@[simp]
+theorem wideSubcategoryInclusion.obj {X} : (wideSubcategoryInclusion P).obj X = X.obj :=
+  rfl
+
+@[simp]
+theorem wideSubcategoryInclusion.map {X Y} {f : X ⟶ Y} :
+    (wideSubcategoryInclusion P ).map f = f.1 :=
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Widesubcategory.lean:110]:
```suggestion
    (wideSubcategoryInclusion P).map f = f.1 :=
```

File: Mathlib/CategoryTheory/Widesubcategory.lean [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+section Induced
+
+variable {C : Type u₁} (D : Type u₂) [Category.{v₁} D]
+variable (F : C → D) (P : MorphismProperty D) [IsMultiplicative P]
+
+/-- `InducedWideCategory D F P`, where `F : C → D`, is a typeclass synonym for `C`,
+which provides a category structure so that the morphisms `X ⟶ Y` are the morphisms
+in `D` from `F X` to `F Y` which satisfy a property `P : MorphismProperty D` that is multiplicative.
+-/
+-- Porting note(#5171): removed @[nolint has_nonempty_instance]
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Widesubcategory.lean:42]:
```suggestion
```

**Comment by joelriou** [comment]
You should edit the title of the PR as "feat(CategoryTheory): ..."

bors d+

**Pull request closed without merging**